<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Source separation - 2&nbsp; Methods</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./references.html" rel="next">
<link href="./00_simulations.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.min.css">

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar floating slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./01_methods.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Methods</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Source separation</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./00_simulations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Simulating hyperspectral fiber photometry signals</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_methods.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Methods</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#baselines" id="toc-baselines" class="nav-link active" data-scroll-target="#baselines"><span class="header-section-number">2.1</span> Baselines</a>
  <ul class="collapse">
  <li><a href="#non-negative-matrix-factorization-nmf" id="toc-non-negative-matrix-factorization-nmf" class="nav-link" data-scroll-target="#non-negative-matrix-factorization-nmf">Non-negative matrix factorization (NMF)</a></li>
  <li><a href="#vanilla-gradient-descent-steps" id="toc-vanilla-gradient-descent-steps" class="nav-link" data-scroll-target="#vanilla-gradient-descent-steps">Vanilla gradient descent steps</a></li>
  <li><a href="#implementation-in-scikit-learn" id="toc-implementation-in-scikit-learn" class="nav-link" data-scroll-target="#implementation-in-scikit-learn">Implementation in scikit-learn</a></li>
  <li><a href="#projected-gradient-descent" id="toc-projected-gradient-descent" class="nav-link" data-scroll-target="#projected-gradient-descent">Projected gradient descent:</a></li>
  <li><a href="#multiplicative-updates" id="toc-multiplicative-updates" class="nav-link" data-scroll-target="#multiplicative-updates">Multiplicative updates</a></li>
  <li><a href="#coordinate-descent" id="toc-coordinate-descent" class="nav-link" data-scroll-target="#coordinate-descent">Coordinate descent</a></li>
  <li><a href="#simple-coordinate-descent" id="toc-simple-coordinate-descent" class="nav-link" data-scroll-target="#simple-coordinate-descent">Simple coordinate descent</a></li>
  <li><a href="#linear-unmixing-approaches" id="toc-linear-unmixing-approaches" class="nav-link" data-scroll-target="#linear-unmixing-approaches">Linear unmixing approaches</a></li>
  </ul></li>
  <li><a href="#training-a-conv-net" id="toc-training-a-conv-net" class="nav-link" data-scroll-target="#training-a-conv-net"><span class="header-section-number">2.2</span> Training a conv-net</a>
  <ul class="collapse">
  <li><a href="#software-tools" id="toc-software-tools" class="nav-link" data-scroll-target="#software-tools">Software tools</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Methods</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><span class="math display">
\gdef\tr{\textrm{tr}}
\gdef\vec{\textrm{vec}}
\gdef\rpos{\mathbb{R_{\geq 0}}}
</span></p>
<p>This chapter documents ideas to analyze the raw signals.</p>
<section id="baselines" class="level2 page-columns page-full" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="baselines"><span class="header-section-number">2.1</span> Baselines</h2>
<section id="non-negative-matrix-factorization-nmf" class="level3">
<h3 class="anchored" data-anchor-id="non-negative-matrix-factorization-nmf">Non-negative matrix factorization (NMF)</h3>
<ul>
<li>Let <span class="math inline">\rpos</span> denote non-negative real number set</li>
<li>Data <span class="math inline">X \in \rpos^{p \times q}</span></li>
<li>Let <span class="math inline">W \in \rpos^{p \times r}</span> and <span class="math inline">H \in \rpos^{r \times q}</span>.</li>
</ul>
<p>The exact NMF problem <span class="math inline">X = WH</span> is NP-hard. The approximate problem <span class="math inline">X \approx WH</span> is non-convex.</p>
</section>
<section id="vanilla-gradient-descent-steps" class="level3">
<h3 class="anchored" data-anchor-id="vanilla-gradient-descent-steps">Vanilla gradient descent steps</h3>
<span class="math display">\begin{aligned}
\frac{\partial{}}{\partial{W}} \bigg[\frac{1}{2} \| Y - AH \|_F^2 \bigg]
    &amp;= \frac{1}{2}\frac{\partial{}}{\partial{W}}\bigg[\tr(Y^TY - Y^TAX - X^TA^TY + X^TA^TAX)   \bigg] \\
    &amp;= \frac{1}{2}\frac{\partial{}}{\partial{W}}\bigg[- \tr(Y^TAX) - \tr(X^TA^TY) + \tr(X^TA^TAX)   \bigg] \\
    
\end{aligned}</span>
</section>
<section id="implementation-in-scikit-learn" class="level3 page-columns page-full">
<h3 class="anchored" data-anchor-id="implementation-in-scikit-learn">Implementation in scikit-learn</h3>
<p><code>scikit-learn</code> provides implementations of the multiplicative update <span class="citation" data-cites="lee1999nmf">Lee and Seung (<a href="references.html#ref-lee1999nmf" role="doc-biblioref">1999</a>)</span> and co-ordinate descent rules to solve the problem, includes regularizers for sparsity and smoothness. The problem considered in scikit-learn is:</p>
<p><span class="math display">
\begin{aligned}
L(W, H) &amp;= \frac{1}{2} \bigg[ \| X - WH \|_F^2 \\
            &amp;+ p \alpha_W \bigg(2f  \|\vec(W)\|_1 + (1 - f) \| W \|_{F}^2\bigg) \\
            &amp;+ q \alpha_H \bigg(2f  \|\vec(H)\|_1 + (1 - f) \| H \|_{F}^2\bigg) \bigg]
\end{aligned}
</span></p>
<ul>
<li><span class="math inline">X \in \mathbb{R_+}^{p \times q}</span></li>
<li><span class="math inline">\vec(A)</span> flattens matrix <span class="math inline">A</span> into a vector</li>
<li><span class="math inline">\alpha_W, \alpha_H , f</span> are input parameters that control regularization</li>
</ul>

<div class="no-row-height column-margin column-container"><div class="">
<p>This sections on projected gradient descent, explanation for multiplicative update, and coordinate descent are reproduced with nominal changes from <a href="https://www.cs.cornell.edu/~bindel/class/sjtu-summer19/lec/2019-05-30.pdf">lecture notes</a> by <a href="https://github.com/dbindel">David Bindel</a>.</p>
</div></div></section>
<section id="projected-gradient-descent" class="level3">
<h3 class="anchored" data-anchor-id="projected-gradient-descent">Projected gradient descent:</h3>
<p>To minimize a function <span class="math inline">\phi(x)</span>, we can use gradient descent. We introduce a projection function <span class="math inline">\mathcal{P}(x)</span> that maps <span class="math inline">x</span> to the nearest feasible point. For the non-negativity constraint of NMF, <span class="math inline">\mathcal{P}(x) = [x]_+</span> is the elementwise maximum of <span class="math inline">x</span> and zero. The projected gradient descent iteration is then <span class="math display">
  x^{k+1} = \mathcal{P}\left( x^{k+1} - \alpha_k \nabla \phi(x^k) \right).
</span></p>
<p>The convergence properties of projected gradient descent are similar to those of the unprojected version.</p>
<p>In order to write the gradient for the NMF objective without descending into a morass of indices, it is helpful to introduce the Frobenius inner product: for matrices <span class="math inline">X, Y \in \mathbb{R}^{m \times n}</span>, <span class="math display">
  \langle X, Y \rangle_F =
  \sum_{i,j} y_{ij} x_{ij} =
  \tr(Y^T X).
</span> The Frobenius inner product is the inner product associated with the Frobenius norm: <span class="math inline">\|X\|_F^2 = \langle X, X \rangle_F</span>, and we can apply the usual product rule for differentiation to compute directional derivatives of <span class="math inline">\phi(W,H) = \|A-WH\|_F^2/2</span> with respect to <span class="math inline">W</span> and <span class="math inline">H</span>: <span class="math display">\begin{align*}
  \delta \phi
  &amp;= \delta \left[ \frac{1}{2} \langle A-WH, A-WH \rangle_F \right] \\
  &amp;= \langle \delta(A-WH), A-WH \rangle_F \\
  &amp;= -\langle (\delta W) H, A-WH \rangle_F
     -\langle W (\delta H), A-WH \rangle_F.
\end{align*}</span> We let <span class="math inline">R = A-WH</span>, and use the fact that the trace of a product of matrices is invariant under cyclic permutations of the matrices: <span class="math display">\begin{align*}
  \langle (\delta W) H, R \rangle_F
  &amp; = \tr(H^T (\delta W)^T R)
    = \tr((\delta W)^T RH^T)
    = \langle \delta W, RH^T \rangle_F \\
  \langle W (\delta H), R \rangle_F
  &amp;= \tr((\delta H)^T W^T R) = \langle \delta H, W^T R \rangle_F.
\end{align*}</span> Therefore, the projected gradient descent iteration for this problem is <span class="math display">\begin{align*}
  W^{\mathrm{new}} &amp;=
  \left[
    W + \alpha RH^T
  \right]_+ \\
  H^{\mathrm{new}} &amp;=
  \left[
    H + \alpha W^T R
  \right]_+,
\end{align*}</span> where in the interest of legibility we have suppressed the iteration index on the right hand side.</p>
</section>
<section id="multiplicative-updates" class="level3">
<h3 class="anchored" data-anchor-id="multiplicative-updates">Multiplicative updates</h3>
<p>One of the earliest and most popular NMF solvers is the multiplicative update scheme of Lee and Seung. This has the form of a scaled gradient descent iteration where we replace the uniform step size <span class="math inline">\alpha_k</span> with a different (non-negative) step size for each entry of <span class="math inline">W</span> and <span class="math inline">H</span>:</p>
<p><span class="math display">
\begin{align*}
  W^{\mathrm{new}} &amp;=
  \left[
    W + S \odot \left( AH^T - W H H^T \right)
  \right]_+ \\
  H^{\mathrm{new}} &amp;=
  \left[
    H + S' \odot \left( W^T A - W^T W H \right)
  \right]_+,  
\end{align*}
</span></p>
<p>where <span class="math inline">\odot</span> denotes elementwise multiplication. We similarly let <span class="math inline">\oslash</span> to denote elementwise division to define the nonnegative scaling matrices</p>
<p><span class="math display">
\begin{align*}
S  = W \oslash (WHH^T) \\
S' = H \oslash (W^TWH)
\end{align*}
</span></p>
<p>With these choices, two of the terms in the summation cancel, so that <span class="math display">
\begin{align*}
  W^{\mathrm{new}} &amp;= S  \odot (AH^T) = W \oslash (WHH^T) \odot (AH^T) \\
  H^{\mathrm{new}} &amp;= S' \odot (W^TA) = H \oslash (W^TWH) \odot (W^TA).
\end{align*}
</span></p>
<p>At each step of the Lee and Seung scheme, we scale the (non-negative) elements of <span class="math inline">W</span> and <span class="math inline">H</span> by non-negative factors, yielding a non-negative result. There is no need for a non-negative projection because the step sizes are chosen increasingly conservatively as elements of <span class="math inline">W</span> and <span class="math inline">H</span> approach zero. But because the steps are very conservative, the Lee and Seung algorithm may require a large number of steps to converge.</p>
</section>
<section id="coordinate-descent" class="level3">
<h3 class="anchored" data-anchor-id="coordinate-descent">Coordinate descent</h3>
<p>The block coordinate descent method (also known as block relaxation, or nonlinear Gauss-Seidel) for solving <span class="math inline">\min~\phi(x_1, x_2, \ldots, x_p)</span> for <span class="math inline">x_i \in \Omega_i</span></p>
<p>involves repeatedly optimizing with respect to one coordinate at a time. In the basic method, we iterate through each <span class="math inline">i</span> and compute <span class="math display">
  x_i^{k+1} =
  \textrm{argmin}_\xi \phi(x_1^{k+1}, \ldots, x_{i-1}^{k+1}, \xi, x_{i+1}^k, \ldots, x_p^k)
</span> The individual subproblems are often simpler than the full problem. If each subproblem has a unique solution (e.g.~if each subproblem is strongly convex), the iteration converges to a stationary point.</p>
</section>
<section id="simple-coordinate-descent" class="level3">
<h3 class="anchored" data-anchor-id="simple-coordinate-descent">Simple coordinate descent</h3>
<p>Perhaps the simplest coordinate descent algorithm for NMF sweeps through all entries of <span class="math inline">W</span> and <span class="math inline">H</span>. Let <span class="math inline">R = A-WH</span>; then for the <span class="math inline">(i,j)</span> coordinate of <span class="math inline">W</span>, we compute the update <span class="math inline">w_{ij} = w_{ij} + s</span> where <span class="math inline">s</span> minimizes the quadratic</p>
<p><span class="math display">
\begin{align*}
\frac{1}{2} \|A-(W+se_i e_j^T) H\|_F^s = \\
\frac{1}{2} \|R\|_F^2 - s \langle (e_i e_j^T) , RH^T \rangle_F + \frac{1}{2} s^2 \|e_i e_j^T H\|_F^2
\end{align*}
</span></p>
<p>subject to the constraint that <span class="math inline">s \geq -w_{ij}</span>. The solution to this optimization is</p>
<p><span class="math display">
  s = \max\left( -w_{ij}, \frac{(RH^T)_{ij}}{(HH^T)_{jj}} \right)
</span></p>
<p>Therefore, the update for <span class="math inline">w_{ij}</span> is <span class="math display">
  s = \max\left( -w_{ij}, \frac{(RH^T)_{ij}}{(HH^T)_{jj}} \right), \quad
  w_{ij} := w_{ij} + s, \quad
  R_{i,:} := R_{i,:} - s H_{j,:}
</span> A similar computation for the elements of <span class="math inline">H</span> gives us the update formulas</p>
<p><span class="math display">
\begin{align*}
  s = \max\left( -h_{ij}, \frac{(W^TR)_{ij}}{(W^TW)_{ii}} \right), \quad
  h_{ij} := h_{ij} + s, \quad
  R_{:,j} := R_{:,j} - s W_{:,i}.
\end{align*}
</span></p>
<p>Superficially, this looks much like projected gradient descent with scaled step lengths. However, where in gradient descent (or the multiplicative updates of Lee and Seung) the updates for all entries of <span class="math inline">W</span> and <span class="math inline">H</span> are independent, in this coordinate descent algorithm we only have independence of updates for a single column of <span class="math inline">W</span> or a single row of <span class="math inline">H</span>. This is a disadvantage for efficient implementation.</p>
</section>
<section id="linear-unmixing-approaches" class="level3">
<h3 class="anchored" data-anchor-id="linear-unmixing-approaches">Linear unmixing approaches</h3>
<p>Review <a href="https://www.nature.com/articles/s41467-022-30168-z">PICASSO</a>.</p>
</section>
</section>
<section id="training-a-conv-net" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="training-a-conv-net"><span class="header-section-number">2.2</span> Training a conv-net</h2>
<ol type="1">
<li>The raw signal is simulated with the generative model in <a href="00_simulations.html#eq-observation">Equation&nbsp;<span>1.8</span></a></li>
<li>A conv-net is trained to reconstruct sensor activity $a(i,t)</li>
<li>Assume <span class="math inline">n(t,j)</span> is given</li>
</ol>
<section id="what-we-learnt" class="level4">
<h4 class="anchored" data-anchor-id="what-we-learnt">What we learnt:</h4>
<ul>
<li>This strategy was implemented for synthetic data, and it works.</li>
<li>The approach does not directly generalize to real data.</li>
<li>The range of parameters used to generate synthetic data are key.</li>
<li>Ideally we’d set these parameters in a rational range, or learn them from data.</li>
<li>Problems in learning parameters for synthetic data:
<ul>
<li>Experimental setup is changing.</li>
<li>Position of notch filters must also be modeled (adds to complexity).</li>
<li>Similarity metric (to compare synthetic with real data) must be well-defined to learn parameters through optimization.</li>
</ul></li>
</ul>
</section>
<section id="software-tools" class="level3">
<h3 class="anchored" data-anchor-id="software-tools">Software tools</h3>
<p><span class="citation" data-cites="simpson2023lights">Simpson et al. (<a href="references.html#ref-simpson2023lights" role="doc-biblioref">2023</a>)</span> is a recent review of best practices in fiber photometry.</p>
<p>A list of software tools and algorithms relevant for fiber photometry data analysis.</p>
<ul>
<li><span class="citation" data-cites="sherathiya2021guppy">Sherathiya et al. (<a href="references.html#ref-sherathiya2021guppy" role="doc-biblioref">2021</a>)</span>:</li>
<li><span class="citation" data-cites="bruno2021pmat">Bruno et al. (<a href="references.html#ref-bruno2021pmat" role="doc-biblioref">2021</a>)</span>:</li>
<li><span class="citation" data-cites="bridge2023fipha">Bridge et al. (<a href="references.html#ref-bridge2023fipha" role="doc-biblioref">2023</a>)</span>:</li>
<li><span class="citation" data-cites="keevers2023obtaining">Keevers, McNally, and Jean-Richard-dit-Bressel (<a href="references.html#ref-keevers2023obtaining" role="doc-biblioref">2023</a>)</span>: Compare of OLS vs.&nbsp;iteratively reweighted least-squares to use isosbestic signal for artifact removal.</li>
<li><span class="citation" data-cites="creamer2022correcting">Creamer et al. (<a href="references.html#ref-creamer2022correcting" role="doc-biblioref">2022</a>)</span>: Motion correction for two channel data, where one of the channels has activity independent signals.</li>
</ul>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-bridge2023fipha" class="csl-entry" role="listitem">
Bridge, Matthew F, Leslie R Wilson, Sambit Panda, Korey D Stevanovic, Ayland C Letsinger, Sandra McBride, and Jesse D Cushman. 2023. <span>“FiPhA: An Open-Source Platform for Fiber Photometry Analysis.”</span> <em>bioRxiv</em>.
</div>
<div id="ref-bruno2021pmat" class="csl-entry" role="listitem">
Bruno, Carissa A, Chris O’Brien, Svetlana Bryant, Jennifer I Mejaes, David J Estrin, Carina Pizzano, and David J Barker. 2021. <span>“pMAT: An Open-Source Software Suite for the Analysis of Fiber Photometry Data.”</span> <em>Pharmacology Biochemistry and Behavior</em> 201: 173093.
</div>
<div id="ref-creamer2022correcting" class="csl-entry" role="listitem">
Creamer, Matthew S, Kevin S Chen, Andrew M Leifer, and Jonathan W Pillow. 2022. <span>“Correcting Motion Induced Fluorescence Artifacts in Two-Channel Neural Imaging.”</span> <em>PLoS Computational Biology</em> 18 (9): e1010421.
</div>
<div id="ref-keevers2023obtaining" class="csl-entry" role="listitem">
Keevers, Luke J, Gavan P McNally, and Philip Jean-Richard-dit-Bressel. 2023. <span>“Obtaining Artifact-Corrected Signals in Fiber Photometry: Isosbestic Signals, Robust Regression and dF/f Calculations.”</span>
</div>
<div id="ref-lee1999nmf" class="csl-entry" role="listitem">
Lee, Daniel D, and H Sebastian Seung. 1999. <span>“Learning the Parts of Objects by Non-Negative Matrix Factorization.”</span> <em>Nature</em> 401 (6755): 788–91.
</div>
<div id="ref-sherathiya2021guppy" class="csl-entry" role="listitem">
Sherathiya, Venus N, Michael D Schaid, Jillian L Seiler, Gabriela C Lopez, and Talia N Lerner. 2021. <span>“GuPPy, a Python Toolbox for the Analysis of Fiber Photometry Data.”</span> <em>Scientific Reports</em> 11 (1): 24212.
</div>
<div id="ref-simpson2023lights" class="csl-entry" role="listitem">
Simpson, Eleanor H, Thomas Akam, Tommaso Patriarchi, Marta Blanco-Pozo, Lauren M Burgeno, Ali Mohebi, Stephanie J Cragg, and Mark E Walton. 2023. <span>“Lights, Fiber, Action! A Primer on in Vivo Fiber Photometry.”</span> <em>Neuron</em>.
</div>
</div>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./00_simulations.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Simulating hyperspectral fiber photometry signals</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./references.html" class="pagination-link">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>